source/adapters/cuda/usm.cpp

```diff
@@ -102,26 +102,47 @@ urUSMSharedAlloc(ur_context_handle_t hContext, ur_device_handle_t hDevice,
   return UR_RESULT_SUCCESS;
 }
 
-ur_result_t USMFreeImpl(ur_context_handle_t, void *Pointer) {
+ur_result_t USMFreeImpl(ur_context_handle_t hContext, void *Pointer) {
   ur_result_t Result = UR_RESULT_SUCCESS;
   try {
     unsigned int IsManaged;
     unsigned int Type;
-    void *AttributeValues[2] = {&IsManaged, &Type};
-    CUpointer_attribute Attributes[2] = {CU_POINTER_ATTRIBUTE_IS_MANAGED,
-                                         CU_POINTER_ATTRIBUTE_MEMORY_TYPE};
-    UR_CHECK_ERROR(cuPointerGetAttributes(2, Attributes, AttributeValues,
-                                          (CUdeviceptr)Pointer));
+    unsigned int DeviceOrdinal;
+    const int NumAttributes = 3;
+    void *AttributeValues[NumAttributes] = {&IsManaged, &Type, &DeviceOrdinal};
+
+    CUpointer_attribute Attributes[NumAttributes] = {
+        CU_POINTER_ATTRIBUTE_IS_MANAGED, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+        CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL};
+    UR_CHECK_ERROR(cuPointerGetAttributes(
+        NumAttributes, Attributes, AttributeValues, (CUdeviceptr)Pointer));
     UR_ASSERT(Type == CU_MEMORYTYPE_DEVICE || Type == CU_MEMORYTYPE_HOST,
               UR_RESULT_ERROR_INVALID_MEM_OBJECT);
-    if (IsManaged || Type == CU_MEMORYTYPE_DEVICE) {
-      // Memory allocated with cuMemAlloc and cuMemAllocManaged must be freed
-      // with cuMemFree
-      UR_CHECK_ERROR(cuMemFree((CUdeviceptr)Pointer));
+
+    std::vector<ur_device_handle_t> Devices = hContext->getDevices();
+    unsigned int NumDevices = Devices.size();
+    UR_ASSERT(DeviceOrdinal < NumDevices, UR_RESULT_ERROR_INVALID_DEVICE);
+
+    ur_device_handle_t Device = Devices[DeviceOrdinal];
```

Comment:
I don't think the CUDA DeviceOrdinal works here because the UR context might not contain all the devices available, using the DeviceOrdinal with the platform might be better, I think the order in the platform is correct but I'm not 100% sure.

Reply:
This is the `ur_result_t USMFreeImpl(ur_context_handle_t hContext, void *Pointer)` function. We have access to `ur_context_handle_t hContext` only here, but AFAIK `ur_context_handle_t` does not have access to the platform, or am I missing something? Is there a way to get the platform from `ur_context_handle_t`?

Reply:
I don't think you can get the platform from the context, but I think you can get it from the device, if that works you could get the platform from any device in the context and then use that to find the correct device.

Reply:
@npmiller Thanks!

Done.